habitat-lab/habitat/tasks/rearrange/rearrange_sensors.py

```diff
@@ -1656,6 +1664,178 @@ def get_observation(self, observations, *args, **kwargs):
+            self.pcl_o3d_list.append(pcl_o3d)
+            if len(self.pcl_o3d_list) > 5:
+                # merge point clouds
+                pcl_o3d = o3d.geometry.PointCloud()
+                for pcl in self.pcl_o3d_list:
+                    pcl_o3d += pcl
+
+                # save point cloud and clear list
+                o3d.io.write_point_cloud("debug_pcl_{}.ply".format(self._debug_save_counter), pcl_o3d)
+                self.pcl_o3d_list = []
+                self._debug_save_counter += 1
+
+        return rgb_obs
+
+
+@registry.register_sensor
+class ArmWorkspaceRGBThirdSensor(ArmWorkspaceRGBSensor):
+    """ Sensor to visualize the reachable workspace of an articulated arm """
+    cls_uuid: str = "arm_workspace_rgb_third"
+
+    def __init__(self, sim, config, *args, **kwargs):
+        super().__init__(sim=sim, config=config, *args, **kwargs)
+
+        self.rgb_sensor_name = config.get("rgb_sensor_name", "third_rgb")
+        self.depth_sensor_name = config.get("depth_sensor_name", "third_depth")
+
+
+@registry.register_sensor
+class ObjectMasksSensor(UsesArticulatedAgentInterface, Sensor):
+    """ Sensor to mask the objects that are interested """
+    cls_uuid: str = "object_masks"
+
+    def __init__(self, sim, config, *args, **kwargs):
+        super().__init__(config=config)
+        self._sim = sim
+
+    def _get_uuid(self, *args, **kwargs):
+        return ObjectMasksSensor.cls_uuid
+
+    def _get_sensor_type(self, *args, **kwargs):
+        return SensorTypes.SEMANTIC
+
+    def _get_observation_space(self, *args, config, **kwargs):
+        # The observation space is flexible, should not be used as gym input
+        return spaces.Box(low=0, high=np.iinfo(np.int64).max, shape=(), dtype=np.int64)
+
+    def _get_objects_info(self, object_ids, object_dict):
+        """ Get the names and handles of objects with given IDs """
+        objects_info = {}
+        for obj_id, obj_dict in zip(object_ids, object_dict.values()):
+            objects_info[obj_id] = obj_dict.full_name
+        return objects_info
+
+    # This method assumes the existence of a method to get the semantic sensor's data
+    def get_observation(self, observations, *args, **kwargs):
+        """ Get the detected objects from the semantic sensor data """
+
+        sg = SceneGraphHSSD()
+        sg.load_gt_scene_graph(self._sim)
+
+        objects_info = self._get_objects_info(self._sim.scene_obj_ids, sg.object_layer.obj_dict)
+
+        semantic_obs = observations['head_semantic'].squeeze()
+        rgb_obs = observations['head_rgb']
+
+        # all_ids = np.unique(semantic_obs)
+        # all_ids = np.delete(all_ids, [0, 1])
+
+        mask = np.isin(semantic_obs, self._sim.scene_obj_ids).astype(np.uint8)
+        # mask = np.isin(semantic_obs, all_ids).astype(np.uint8)
+        colored_mask = np.zeros_like(rgb_obs)
+        colored_mask[mask == 1] = [0, 255, 0]
+        masked_rgb = cv2.addWeighted(rgb_obs, 0.7, colored_mask, 0.3, 0)
```

Comment:
Can you set a flag or assertion to check the scene is from the hssd dataset? Since each dataset has different scene graph class to customize initialization process. Using SceneGraphHSSD without extra checking isn't safe.

Reply:
Solved.